ordering/__init__.py

```diff
@@ -68,11 +68,13 @@ def __contains__(self, item: _T) -> bool:
         return item in self._labels
 
     def __iter__(self) -> Iterator[T]:
-        item = self._successors[self._start]
-
-        while not isinstance(item, _Sentinel):
+        successors = self._successors
+        item = successors[self._start]
+        end = self._end
+        while item is not end:
+            assert not isinstance(item, _Sentinel)
             yield item
-            item = self._successors[item]
+            item = successors[item]
```

Comment:
I'm becoming increasingly convinced that rewriting in the style of `OrderedDict` is the right approach.

Though I do wonder, why did you put double underscores on `__start` and `__end` in your example?